run-tests.py

```diff
@@ -0,0 +1,145 @@
+    run_all (bool): Whether to run tests for all projects. If False, only tests for changed directories are run.
+    verbosity (Optional[str]): Verbosity level. If None, nothing is logged. If "v", step description is logged. If "vv",
+        pytest output and detailed information are logged.
+    RETURN (int): Return code.
+    """
+    logger = _get_logger()
+    cmds_status: List[Dict[str, Union[str, int, List[str]]]] = []
+    top_level_dirs = ("benchmarks", "experimental", "integrations", "pipelines", "tutorials")
+
+    _log(logger, verbosity, "v", "Scanning projects")
+    if run_all:
+        proj_dirs = [
+            path for path in
+            [Path(tl_dir) / subdir for tl_dir in top_level_dirs for subdir in os.listdir(tl_dir)]
+            if path.is_dir()
+        ]
+    else:
+        # Fetch which files were changed in the last commit.
+        _run(["git", "diff", "--name-only", "HEAD", "HEAD~1"], cmds_status)
+        proj_dirs = {
+            Path(file).parent for file in
+            _run(
+                ["git", "diff", "--name-only", "HEAD", "HEAD~1"], cmds_status
+            ).stdout.decode("utf-8").split("\n")
+            if Path(file).parent != Path(".")
+        }
+
+    for proj_dir in proj_dirs:
+        _log(logger, verbosity, "v", f"*** {proj_dir} ***")
+
+        # Install from requirements.txt, if it exists.
+        if (proj_dir / "requirements.txt").exists():
+            _log(logger, verbosity, "v", "  - Installing requirements")
+            _run([python, "-m", "pip", "-q", "install", "-r", proj_dir / "requirements.txt"], cmds_status)
+
+        # Fetch spacy version from project.yml, install spacy version if available.
+        spacy_version = srsly.read_yaml(proj_dir / "project.yml").get("spacy_version")
+        if spacy_version:
+            _log(logger, verbosity, "v", f"  - Installing spacy{spacy_version} from project.yml")
+            _run([python, "-m", "pip", "-q", "install", f"spacy{spacy_version}", "--force-reinstall"], cmds_status)
+
+        _log(logger, verbosity, "v", "  - Running tests")
+        _run([python, "-m", "pytest", "-q", "-s", proj_dir], cmds_status)
+        # Log stderr and stdout.
+        if cmds_status[-1]["returncode"] not in (0, 5):
+            for channel in ("stderr", "stdout"):
+                _log(logger, verbosity, "vv", f"    ### {channel} ###")
+                for out in cmds_status[-1][channel]:
+                    _log(logger, verbosity, "vv", f"    {out}")
+        elif cmds_status[-1]["returncode"] == 5:
+            _log(logger, verbosity, "vv", "    No tests found")
+
+        _log(logger, verbosity, "v", "  - Restoring environment")
+        with tempfile.NamedTemporaryFile("w") as file:
+            file.writelines(
+                _run(
+                    [python, "-m", "pip", "freeze", "--exclude", "torch", "cupy-cuda111"], cmds_status
+                ).stdout.decode("utf-8").split("\n")
+            )
```

Comment:
`writelines` does not add line separators, so the input here is run together and the resulting file is one line. 

```suggestion
            file.write(
                _run(
                    [python, "-m", "pip", "freeze", "--exclude", "torch", "cupy-cuda111"], cmds_status
                ).stdout.decode("utf-8")
            )
```